tests/src/main/java/com/hazelcast/stabilizer/tests/map/MapHeapHogTest.java

```diff
@@ -1,116 +1,149 @@
         if (isMemberNode(targetInstance)) {
-            while (targetInstance.getCluster().getMembers().size() != 3) {
-                log.info(basename + " waiting cluster == 3");
-                Thread.sleep(1000);
+            while (targetInstance.getCluster().getMembers().size() != MIN_CLUSTER_SIZE) {
+                LOGGER.info(basename + " waiting to hit cluster size " + MIN_CLUSTER_SIZE);
+                sleepSeconds(1);
             }
         }
     }
 
-    @Run
-    public void run() {
-        ThreadSpawner spawner = new ThreadSpawner(testContext.getTestId());
-        for (int k = 0; k < threadCount; k++) {
-            spawner.spawn(new Worker());
+    @Verify(global = false)
+    public void verify() {
+        if (isMemberNode(targetInstance)) {
+            StringBuilder sb = new StringBuilder("Local verify of ").append(basename).append("\nmap size = ").append(map.size());
+            addMemoryStatistics(sb);
+            LOGGER.info(sb.toString());
         }
-        spawner.awaitCompletion();
     }
 
+    @RunWithWorker
+    public AbstractWorkerTask createWorker() {
+        return new Worker();
+    }
+
+    private class Worker extends AbstractMonotonicWorkerTask {
+        private final StringBuilder sb = new StringBuilder();
+        private final long localStartKey = (useSameKeysOnAllWorker ? 0 : startKey.getAndIncrement());
 
-    private class Worker implements Runnable {
         @Override
-        public void run() {
-            while (!testContext.isStopped()) {
-                long free = Runtime.getRuntime().freeMemory();
-                long total = Runtime.getRuntime().totalMemory();
-                long used = total - free;
-                long max = Runtime.getRuntime().maxMemory();
-                long totalFree = max - used;
-
-                long maxLocalEntries = (long) ((totalFree / approxEntryBytesSize) * approxHeapUsageFactor);
-
-                long key = 0;
-                for (int i = 0; i < maxLocalEntries; i++) {
-                    key = nextKeyOwnedBy(key, targetInstance);
-                    map.put(key, key, ttlHours, TimeUnit.HOURS);
-                    key++;
+        public void timeStep() {
+            long free = Runtime.getRuntime().freeMemory();
+            long total = Runtime.getRuntime().totalMemory();
+            long max = Runtime.getRuntime().maxMemory();
+
+            long used = total - free;
+            long totalFree = max - used;
+
+            long maxLocalEntries = (long) ((totalFree / APPROX_ENTRY_BYTES_SIZE) * approxHeapUsageFactor);
+            long maxIterations = (long) ((double) maxLocalEntries / (double) threadCount);
+
+            long key = localStartKey;
+            for (int workerIteration = 0; workerIteration < maxIterations; workerIteration++) {
+                key = nextKeyOwnedBy(targetInstance, key, (useSameKeysOnAllWorker ? 1 : threadCount));
+                map.put(key, key, ttlHours, TimeUnit.HOURS);
```

Comment:
This makes no sense. You just create a huge race between multiple worker threads, which will all try to insert the same key. The `putIfAbsent` will not change anything, but increase the load on the cluster, due to the additional checks.